massa-factory-worker/src/denunciation_factory.rs

```diff
@@ -0,0 +1,260 @@
+
+    // internal for block header (or secured header)
+    /// Internal storage for endorsement -
+    /// store at max 1 endorsement per entry, as soon as we have 2 we produce a Denunciation
+    // FIXME: Store per 'Slot' then per 'PubKey'
+    block_header_by_slot: HashMap<Slot, Vec<SecuredHeader>>,
+}
+
+impl DenunciationFactoryWorker {
+    /// Creates the `FactoryThread` structure to gather all data and references
+    /// needed by the factory worker thread.
+    pub(crate) fn spawn(
+        cfg: FactoryConfig,
+        channels: FactoryChannels,
+        factory_receiver: Receiver<()>,
+        consensus_receiver: Receiver<SecuredHeader>,
+        endorsement_pool_receiver: Receiver<SecureShareEndorsement>,
+    ) -> thread::JoinHandle<()> {
+        thread::Builder::new()
+            .name("denunciation-factory".into())
+            .spawn(|| {
+                let mut factory = Self {
+                    cfg,
+                    channels,
+                    factory_receiver,
+                    consensus_receiver,
+                    endorsement_pool_receiver,
+                    endorsements_by_slot_index: Default::default(),
+                    block_header_by_slot: Default::default(),
+                };
+                factory.run();
+            })
+            .expect("failed to spawn thread : denunciation-factory")
+    }
+
+    /// Process new secured header (~ block header)
+    fn process_new_secured_header(&mut self, secured_header: SecuredHeader) {
+        let key = secured_header.content.slot;
+
+        if self.block_header_by_slot.len() > DENUNCIATION_FACTORY_BLOCK_HEADER_CACHE_MAX_LEN {
+            warn!(
+                "Denunciation factory cannot process - cache full: {}",
+                self.block_header_by_slot.len()
+            );
+            return;
+        }
+
+        let denunciation_: Option<Denunciation> = match self.block_header_by_slot.entry(key) {
+            Entry::Occupied(mut eo) => {
+                let secured_headers = eo.get_mut();
+                // Store at max 2 SecuredHeader's
+                if secured_headers.len() == 1 {
+                    secured_headers.push(secured_header);
+
+                    // TODO / OPTIM?: use [] ?
+                    // safe to unwrap as we just checked the vec len
+                    let header_1 = secured_headers.get(0).unwrap();
+                    let header_2 = secured_headers.get(1).unwrap();
+                    let de_ = Denunciation::try_from((header_1, header_2));
+                    if let Err(e) = de_ {
+                        debug!("Denunciation factory cannot create denunciation from block headers: {}", e);
+                        return;
+                    }
+                    Some(de_.unwrap()) // Already checked for error
+                } else {
+                    // Already 2 entries - so a Denunciation has already been created
```

Comment:
possible reach here if secured_headers.len() == 0 ? (don't know if possible to have len = 0)

Reply:
It's not even possible because there is a match against Entry::Vacant